File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/inheritance/SubclassTest.scala
package com.github.debop4s.data.tests.mapping.inheritance

import com.github.debop4s.core.utils.Hashs
import com.github.debop4s.data.model.LongEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.util.Date
import javax.persistence._
import org.hibernate.{annotations => hba}
import org.junit.Test


/**
 * subclass 로 구현한 상속 클래스의 속성은 모두 nullable 이어야 한다.
 * (한 테이블에 다른 클래스의 속성이 들어갈 때 현 클래스의 속성은 nullable 로 설정되어야 한다.)
 * Created by debop on 2014. 3. 9.
 */
@org.springframework.transaction.annotation.Transactional
class SubclassTest extends AbstractJpaTest {

  @PersistenceContext val em: EntityManager = null

  @Test
  def subclass() {
    val bankAccount = new SubclassBankAccount()
    bankAccount.account = "account"
    bankAccount.bankname = "bank name"
    bankAccount.owner = "debop"
    em.persist(bankAccount)

    val creditCard = new SubclassCreditCard()
    creditCard.number = "1111-1111-1111-1111"
    creditCard.companyName = "카드사"
    creditCard.expYear = 2020
    creditCard.expMonth = 12
    creditCard.owner = "debop"
    em.persist(creditCard)

    em.flush()
    em.clear()

    val bankAccount1 = em.find(classOf[SubclassBankAccount], bankAccount.id)
    assert(bankAccount1 == bankAccount)

    val creditCard1 = em.find(classOf[SubclassCreditCard], creditCard.id)
    assert(creditCard1 == creditCard)

    em.remove(bankAccount1)
    em.remove(creditCard1)
    em.flush()
    em.clear()

    assert(em.find(classOf[SubclassBankAccount], bankAccount.id) == null)
    assert(em.find(classOf[SubclassCreditCard], creditCard.id) == null)
  }

}

/**
 *  한 테이블에 Super-Sub class 들 모두 저장됩니다. (subclass)
 */
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@hba.DynamicInsert
@hba.DynamicUpdate
abstract class SubclassBillingBase extends LongEntity {

  @Column(nullable = false)
  @Index(name = "ix_subclass_billing_owner", columnList = "owner")
  var owner: String = _

  override def hashCode(): Int = Hashs.compute(owner)
}

@Entity
@hba.DynamicInsert
@hba.DynamicUpdate
class SubclassBankAccount extends SubclassBillingBase {

  var account: String = _
  var bankname: String = _
  var swift: String = _

  override def hashCode(): Int = Hashs.compute(owner, account)
}

@Entity
@hba.Cache(region = "inheritance", usage = hba.CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
class SubclassCreditCard extends SubclassBillingBase {

  var companyName: String = _
  var number: String = _
  var expMonth: Integer = _
  var expYear: Integer = _

  @Temporal(TemporalType.DATE)
  var startDate: Date = _

  @Temporal(TemporalType.DATE)
  var endDate: Date = _

  var swift: String = _

  override def hashCode(): Int = Hashs.compute(owner, number)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/inheritance/subclass/SubclassTest.java
package debop4s.data.orm.jtests.mapping.inheritance.subclass;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.inheritance.subclass.SubclassJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 22. 오후 5:42
 */
@Slf4j
@Transactional
public class SubclassTest extends JpaTestBase {

    @PersistenceContext
    EntityManager em;

    @Test
    public void subclassEntity() {
        SubclassBankAccount bankAccount = new SubclassBankAccount();
        bankAccount.setAccount("account");
        bankAccount.setBankname("은행이름");
        bankAccount.setOwner("배성혁");

        SubclassCreditCard creditCard = new SubclassCreditCard();
        creditCard.setNumber("1111-1111-1111-1111");
        creditCard.setExpYear(2020);
        creditCard.setExpMonth(12);
        creditCard.setOwner("배성혁");

        em.persist(bankAccount);
        em.persist(creditCard);
        em.flush();

        em.clear();

        SubclassBankAccount bankAccount1 = em.find(SubclassBankAccount.class, bankAccount.getId());
        assertThat(bankAccount1).isEqualTo(bankAccount);

        SubclassCreditCard creditCard1 = em.find(SubclassCreditCard.class, creditCard.getId());
        assertThat(creditCard1).isEqualTo(creditCard);

        em.remove(bankAccount1);
        em.remove(creditCard1);

        assertThat(em.find(SubclassBankAccount.class, bankAccount.getId())).isNull();
        assertThat(em.find(SubclassCreditCard.class, creditCard.getId())).isNull();
    }

}
